Extend Multiselect: Select All component

Lelith opened the pull request:
# Purpose of PR

This extends the Multiselect component for another compount component "SelectAll". This component renders another MultiSelect Option that expects a callback function which will select all your elements. 

<img width="110" alt="image" src="https://user-images.githubusercontent.com/1789174/205282150-56586ad7-f086-45a7-ab5e-b8cf7cbf0f45.png">

<img width="136" alt="image" src="https://user-images.githubusercontent.com/1789174/205282359-41dc06a3-d9da-44e5-9cee-3dc17c956e44.png">

How it works:
Add the SelectAll component somewhere in your Multiselect List. Then create a function that will toggle all your available options, just as any other onSelectItem handler, it expects an event as property.
In order to recognize when all options have been selected, it is recommended to implement another function for that, which uses a useMemo hook to detect changes in the current selection, and then check if all available options are selected.
See also the Storybook and documentation for example implementations.

Also fixes some minor issues:
- design of storybook example
- makes the width dependent of size of surrounding container
- the element list has now the fixed size of the trigger Button
- the 'n more' text is now in gray600
- adjusted the size of the search bar to be small

bgutsol reviewed (commented):
Huge work @Lelith  💪
I can see that because `Multiselect` is a controlled component you had to introduce a lot of complexity and hacky logic for that `checkAll` functionality. Have you considered just adding an example for the users where they can easily build it on their own? And maybe just provide some compound component for styling purposes. Seems like it will reduce complexity, won't reduce performance (no need to call every option callback to check all of them), and leaves flexibility to the end user (in case they want to adjust label or tweak a bit "select all" logic or pass a tracking on the callback). WDYT?
[For example](https://f36.contentful.com/playground?code=N4Igxg9gJgpiBcICWBbADhATgFwAQCUYBDMPAM0whVwHJNjSaBuAHQDtUMddhcBlbCQDWuAL64KVWgAFIbbDHlkArgBsA9GQDMANgC0kdBDaLsAZ2btOWPLwCya7EjMxVMUmImVqNWcYVKapq6eiiOzq7u2JZs7DAAHlx4sGREjhLKbKRIxrgOqk4ubqQAQkRmSGAAqmZEAOYwAKLxROhuABQAlDzsuLhyZnhmaCQwZrgAvLgA2r19tAAqmEQAbq64JaoQdTQANHN9NABiSGxsRFkwG1s7+2zztCfYJmbjAIJoaHsHtAByMAB3cYAdRgACMKgpvvd5jQYABhKgoGCYMBXeFEQQ3ZQwaEPGgABQAFhBsBBcABxIiqNyYACeeNwAF1WLEYep1LgFkTnLgAAbKFwCTEwPm4Xl1CCnOq4Mm4QZYK7YIlXIpRGBQXAsEDDUba+XkgFXMAXeUkgHivCnWUq3BVACScwGeGmatIGr4IzRZl28pg2D4kXdUE9ozMTMmBAY2AAdIKYMKFO1pkzOqy+s6ZtTVIHigooL6XNg3jTc+qoBGpoQSLH44mYO1UqoXGn2HMOVyeeMVFknLkAUgabgwcbsxrcMY0TarvHMH68+NTUgFNRrcqrqpnHgIGQJ2g+2wzE7jINcESLlA3GXSPaV5H2jA1vJuhMAHw9GEZk+2foqsBCDVfRWakcU8KZH1MGNBEwBpsHTeYkF3dowD-ACoG6YAfj6Itr3zUNvXado0HoFZ6xfd9phjKjiMfesgJAmBU3gvpxFcFwPweL9DzwEwAVwj0vTGSM3TwwSzBjMhBwUTBCN1NFyPlQTcAAQgmKZgNUHFW0-eYcKDUSw3aXj+JDMTtIeUQ5lEVlj24+cohLVR7wg59JnfTCdMzXgUPcNCwNwFzY2g2DmL9YtS30jVkNQjVzIQpCfP-WKOM4vS8wEwy5LGToflY5srg8ziwpM-CxmTVNcqstsYXobBlEwe52h+AAeYV-wkNx4gAESQehsmMCZtUgTSUFiEBcGpJA6jYW8YBQMxBp1aDsG1V8sNwZr8kKfT1ozer6HkXCcjYCZgBEjLvUsnSHgwDA1kwAlKDQBbgF4ZwjjUVRgSQKBlXgWVMFA0QrqKtbrvmTbwhEmMAHl92O3aistOb7SgRbcMc7VEaKjScUW7MsfBpHcFUIgR1UU7sxM3AAH5aC6sZ9ImmkaFwf6aFw5nVBoEHOJYFa8GRlBUfR-TMZAdb+f5vpjFw2aUFOkTHN5h4pcF5x4RitHgCpyKoBV4ncHUMHibOsSYxQIg0FkwSFMKw3Mw04TBKgiAABkICNTAMRcLoY3oNBSbRdp1BYFgzHUOpfRoPQaDi4navqxrsYhraIjzWH4eMFOHgAulTr5PO9AAEmADTgb5A3DaFkXgD5LKS7L6kK6rw3cZgRXBNb4nSfJzvRm7pHZf0+XTvPNhLwTEeV0HoqNa1xW9dK8TTjATTYDMdoss6Wf5mNlP4840Qd-W5r1DTkSTYh9Q2qEK-tKukBRCAA)

Chaoste commented on packages/components/multiselect/README.mdx:
```diff
@@ -68,6 +68,14 @@ To enable searching the following properties have to be set.
 
+### SelectAll shortcut
+
+To offer a shortcut for selecting and deselecting all options, use the property hasSelectAll. This will render a checkbox on top of the list which will trigger the selection and deselection of all elements.
```
Here you say "hasSelectAll" but in the code it's "hasCheckAll"

Chaoste commented on packages/components/multiselect/src/Multiselect.tsx:
```diff
@@ -60,6 +61,12 @@ export interface MultiselectProps extends CommonProps {
   searchInputName?: string;
 
+  /**
+   * Enables a select All shortcut checkbox
```
```suggestion
   * Enables a shortcut checkbox to select or delect all items
```

Chaoste commented on packages/components/multiselect/src/SelectAllOption.tsx:
```diff
@@ -0,0 +1,91 @@
+  }
+};
+
+const areChildrenAllSelected = (
```
```suggestion
const areAllChildrenSelected = (
```

Chaoste commented on packages/components/multiselect/src/SelectAllOption.tsx:
```diff
@@ -0,0 +1,91 @@
+  }
+};
+
+const areChildrenAllSelected = (
```
Also, this looks like copied code from `Multiselect`. If there is time, I think it would b worth putting this into a util file and just import it twice

Lelith replied:
It is very similar to the "iterateOverChildren" functionality which we do to count them. I think i can find a abstraction so we can put that function into a utils file and probaply also merge it with the enriching function. All of them are recursively checking the child components and then do different things with the found options.

Chaoste replied:
True, if the generalization is too complex, we can easily skip this. No need to write super intelligent utilities for the sake of DRY :)

Lelith replied:
the only thing they have in common is that they work recursively over the children. which is neccesary because we don t have power over the child components in this component. which gives us some advantags as well as some problems.
maybe this would be the point to question if it is better to have the children as compound components or if we should take the render of the children in our own hands. but that leads then to the same limitations that the autocomplete has.

i think there is no great answer, but i also dislike the high complexity of the solution.

Chaoste commented on packages/components/multiselect/src/SelectAllOption.tsx:
```diff
@@ -0,0 +1,91 @@
+  );
+
+  // checks if all children are true
+  React.useEffect(() => {
+    const allTrue = areChildrenAllSelected(
+      childNodes,
+      (child) => child.type === MultiselectOption,
+    );
+    setAllSelected(allTrue);
+  }, [childNodes]);
```
I might be missing a obstacle here but I would try to avoid these kinds of state updates. You listen to a (transformed version) of an input variable and then set a state variable. I think that a `UseMemo` would be more appropriate since it saves you one render cycle. Here you first have one cycle with the new prop and then another one to update the state.

Chaoste replied:
If I'm not mistaken than this effet will get called when you change the `checked` state for a checkbox. This means that you don't need the state variable at all and can also save the call at line 73. By just updating the DOM nodes, this component rerenders and updates the memorized variable `allSelected` (instead of a controlled state)

Chaoste replied:
I might be wrong here because Multiselect is controlled but just wanted to raise the possibility of reducing complexity. Dunno if it's possible.

Lelith replied:
i will have another look at it, but without the effect, it has some weird double-clicking required issues.

Chaoste replied:
I think it's also not a problem to have it this way. It's rather a micro-optimization that doesn't need to block this PR :)

Lelith replied:
So i took another go at it and i don't need to call the method when initializing, its enough with the effect. In this case childNodes are props that change, not another state variable or anything inside. So useEffect is the correct hook over useCallback or useMemo.

Chaoste reviewed (commented):
I think this is a nice feature that saves us from writing the same logic multiple times. Thanks for implementing :)

denkristoffer commented:
I agree with @bgutsol that this is a lot to go against the component's implementation. Is the problem here that the MultiSelect is controlled only? This seems to implement a sort of in-between, where we have an uncontrolled part of a controlled component. Am I misunderstanding that?

Lelith commented:
I totally get that this becomes overly complex, and shuts down a part of logic from the user. I am not happy with it either.

The main issue when implementing outside of the component is that you also have to recognise somehow when all options are already selected. For our story example that is quite simple done when you compare the list of available options with the list of selected options.

But when we go in the actual use case, there is most likely a form controller there, too. that has initially selected items e.g. or items that are not longer part of the available options, but are still selected, coming from the saved form.
So you have to then write yourself a function that is identifying which options are there, and if they are selected, and if the current selection state reflects one of "all elements selected"

Also setting a selection, in the real world, is not only done by adding it to a state, that is only a shallow copy of the available options. For having it selected in a controlled form, you need to trigger a real onChange handler, with a real event. So you still need to loop trough all options and call an onChange event for them.

You also need to implement it for every single multi-select which shall have the shortcut.

Having there the relation of children is kind of simpler, even tough with the possibility of having nested children, means we have to loop trough them recursively. 

But i will provide an outside implementation example and then we can see which one is suiting more, and maybe i can put the overly complex one in experience components?

Lelith commented:
@denkristoffer @bgutsol i refactored it and removed the complex logic, to only have a compund component and let the user decide themselve how they want to deal with toggling and detecting and updated storybook and the example accordingly.

I think ultimately delivering the lightweight solution for the majority of users will be better. And if i find myself fighting in a corner too much in the apps, i still can create a "with Select All" HOC

bgutsol commented on packages/components/multiselect/src/Multiselect.tsx:
```diff
@@ -60,6 +60,12 @@ export interface MultiselectProps extends CommonProps {
    */
   searchInputName?: string;
 
+  /**
+   * Enables a shortcut checkbox to select or delect all items
+   * @default false
+   */
+  hasCheckAll?: boolean;
```
don't need this anymore

bgutsol commented on packages/components/multiselect/src/Multiselect.tsx:
```diff
@@ -136,7 +142,7 @@ const countMatchingChildren = (
 function _Multiselect(props: MultiselectProps, ref: React.Ref<HTMLDivElement>) {
   const {
-    className,
+    className = '',
```
Do we really need it?

Lelith replied:
oh that was just an experiment i forgot to remove again. thanks

bgutsol commented on packages/components/multiselect/src/SelectAllOption.tsx:
```diff
@@ -0,0 +1,31 @@
+      value="all"
+      label={displayLabel}
+      itemId="SelectAll"
+      onSelectItem={(event) => onSelectItem(event)}
```
Can we just pass a callback directly?
```suggestion
      onSelectItem={onSelectItem}
```

bgutsol commented on packages/components/multiselect/stories/Multiselect.stories.tsx:
```diff
@@ -95,21 +95,20 @@ export const Basic = () => {
-        (fruit) => fruit !== currentFruit.name,
-      );
-      setSelectedFruits(newSelectedFruits);
+      //its important to use prevState to avoid race conditions when using the state variable as reference.
```
nice 👍

Chaoste replied:
That's definitely a good catch!

bgutsol reviewed (approved):
Thanks for addressing our comments ❤️
only some small nits

Chaoste commented on packages/components/multiselect/src/Multiselect.test.tsx:
```diff
@@ -249,3 +256,22 @@ describe('Multiselect with search', () => {
+    await user.click(
+      screen.getByRole('button', { name: 'Toggle Multiselect' }),
+    );
+    expect(
+      screen.queryByRole('checkbox', { name: 'Deselect all' }),
+    ).toBeInTheDocument();
+    await user.click(screen.queryByRole('checkbox', { name: 'Deselect all' }));
```
```suggestion
    expect(
      screen.getByRole('checkbox', { name: 'Deselect all' }),
    ).toBeInTheDocument();
    await user.click(screen.getByRole('checkbox', { name: 'Deselect all' }));
```
